File: back/src/main/java/com/openclassrooms/mddapi/services/UserService.java
package com.openclassrooms.mddapi.services;

import com.openclassrooms.mddapi.model.dto.UserDTO;

/**
 * Service définissant les opérations métier liées à la gestion des utilisateurs.
 * <p>
 * Cette interface regroupe l’ensemble des fonctionnalités permettant de :
 * <ul>
 *     <li>Créer, mettre à jour et supprimer un utilisateur,</li>
 *     <li>Récupérer un utilisateur par son identifiant, son e-mail ou son nom d’utilisateur,</li>
 *     <li>Vérifier l’existence d’un utilisateur,</li>
 *     <li>Gérer les thèmes suivis par un utilisateur (ajout ou suppression).</li>
 * </ul>
 * </p>
 *
 * <p>
 * L’implémentation par défaut de cette interface est
 * {@link com.openclassrooms.mddapi.services.impl.UserServiceImpl}.
 * </p>
 *
 * @see com.openclassrooms.mddapi.model.dto.UserDTO
 * @see com.openclassrooms.mddapi.model.entities.User
 *
 * @author PA-SCHAMING
 * @version 1.0
 */
public interface UserService {

    /**
     * Recherche un utilisateur à partir de son adresse e-mail.
     *
     * @param email l’adresse e-mail recherchée.
     * @return le {@link UserDTO} correspondant, ou {@code null} si aucun utilisateur n’existe avec cet e-mail.
     */
    UserDTO getUserByEmail(String email);

    /**
     * Recherche un utilisateur à partir de son nom d’utilisateur (username).
     *
     * @param username le nom d’utilisateur recherché.
     * @return le {@link UserDTO} correspondant, ou {@code null} si aucun utilisateur n’existe avec ce nom.
     */
    UserDTO getUserByUsername(String username);

    /**
     * Crée un nouvel utilisateur dans la base de données.
     *
     * @param userDto l’objet {@link UserDTO} représentant le nouvel utilisateur.
     * @return le {@link UserDTO} sauvegardé (sans le mot de passe).
     */
    UserDTO saveUser(UserDTO userDto);

    /**
     * Récupère un utilisateur à partir de son identifiant unique.
     *
     * @param id l’identifiant de l’utilisateur.
     * @return le {@link UserDTO} correspondant.
     */
    UserDTO getUserById(int id);

    /**
     * Supprime un utilisateur à partir de son identifiant.
     *
     * @param id l’identifiant de l’utilisateur à supprimer.
     */
    void deleteUserById(int id);

    /**
     * Met à jour les informations d’un utilisateur existant.
     *
     * @param userDto l’objet {@link UserDTO} contenant les nouvelles informations.
     * @return le {@link UserDTO} mis à jour (sans le mot de passe).
     */
    UserDTO updateUser(UserDTO userDto);

    /**
     * Vérifie si un utilisateur existe à partir de son identifiant.
     *
     * @param id l’identifiant à vérifier.
     * @return {@code true} si l’utilisateur existe, sinon {@code false}.
     */
    boolean checkId(Integer id);

    /**
     * Ajoute un thème à la liste des thèmes suivis par un utilisateur.
     *
     * @param user    l’utilisateur concerné.
     * @param themeId l’identifiant du thème à suivre.
     * @return le {@link UserDTO} mis à jour.
     */
    UserDTO addThemeToFollowed(UserDTO user, int themeId);

    /**
     * Retire un thème de la liste des thèmes suivis par un utilisateur.
     *
     * @param user    l’utilisateur concerné.
     * @param themeId l’identifiant du thème à ne plus suivre.
     * @return le {@link UserDTO} mis à jour.
     */
    UserDTO removeThemeToFollowed(UserDTO user, int themeId);
}

File: front/src/app/shared/services/user.service.ts
import {HttpClient} from "@angular/common/http";
import {Injectable} from "@angular/core";
import {Observable} from "rxjs";
import {User} from "../../mdd/models/user.model";
import {environment} from "../../../environments/environment";
import {Theme} from "../../mdd/models/theme.model";

@Injectable()
/**
 * Service responsable de la gestion des utilisateurs.
 *
 * Fonctionnalités :
 * - Récupérer un utilisateur par son identifiant
 * - Récupérer les thèmes suivis par l'utilisateur courant
 * - Mettre à jour les informations d'un utilisateur
 *
 * Toutes les opérations s'effectuent via les endpoints du backend sous `/api/user`.
 */
export class UserService {

  /**
   * @param httpClient Client HTTP utilisé pour communiquer avec l’API backend.
   */
  constructor(private httpClient: HttpClient) {}

  /**
   * Récupère un utilisateur en fonction de son identifiant.
   *
   * Endpoint :
   * `GET /api/user/:id`
   *
   * @param id Identifiant de l'utilisateur.
   * @returns Observable émettant l'utilisateur correspondant.
   */
  getUserById(id: number): Observable<User> {
    return this.httpClient.get<User>(`${environment.apiUrl}/api/user/${id}`);
  }

  /**
   * Récupère la liste des thèmes suivis par l'utilisateur connecté.
   *
   * Endpoint :
   * `GET /api/user/followed`
   *
   * @returns Observable émettant la liste des thèmes suivis.
   */
  getFollowedThemes(): Observable<Theme[]> {
    return this.httpClient.get<Theme[]>(`${environment.apiUrl}/api/user/followed`);
  }

  /**
   * Met à jour les informations d'un utilisateur.
   *
   * Endpoint :
   * `PUT /api/user/update`
   *
   * @param user Objet utilisateur contenant les nouvelles données.
   * @returns Observable émettant l'utilisateur mis à jour.
   */
  updateUser(user: User): Observable<User> {
    return this.httpClient.put<User>(`${environment.apiUrl}/api/user/update`, user);
  }
}
